File: server/routes/renewals.js
/*const express = require("express");
const router = express.Router();
const pool = require("../config/db");

// GET all renewals
router.get("/", async (req, res) => {
  try {
    const result = await pool.query(
      "SELECT * FROM renewals ORDER BY id ASC"
    );
    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Server Error" });
  }
});

// POST a new renewal
router.post("/", async (req, res) => {
  try {
    const {
      contract_id,
      renewal_date,
      reminder_days,
      status,
    } = req.body;

    /*const result = await pool.query(
      `INSERT INTO renewals
      (contract_id, renewal_date, reminder_days, status)
      VALUES ($1, $2, $3, $4)
      RETURNING *`,
      [contract_id, renewal_date, reminder_days, status]
    );






const result = await pool.query(`
  SELECT
    renewals.id,
    contracts.contract_name,
    contracts.client_name,
    renewals.renewal_date,
    renewals.reminder_days,
    renewals.status
  FROM renewals
  JOIN contracts
    ON renewals.contract_id = contracts.id
  ORDER BY renewals.id ASC
`);




    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Server Error" });
  }
});

module.exports = router; */





/*
// GET all renewals
router.get("/", async (req, res) => {
  try {
    const result = await pool.query(`
      SELECT
        renewals.id,
        contracts.contract_name,
        contracts.client_name,
        renewals.renewal_date,
        renewals.reminder_days,
        renewals.status
      FROM renewals
      JOIN contracts
      ON renewals.contract_id = contracts.id
      ORDER BY renewals.id ASC
    `);

    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Server Error" });
  }
});*/



const express = require("express");
const router = express.Router();
const pool = require("../config/db");

// GET all renewals
router.get("/", async (req, res) => {
  try {
    const result = await pool.query(`
      SELECT
        renewals.id,
        contracts.contract_name,
        contracts.client_name,
        renewals.renewal_type,
        renewals.renewal_date,
        renewals.reminder_days,
        renewals.status
      FROM renewals
      JOIN contracts
        ON renewals.contract_id = contracts.id
      ORDER BY renewals.id ASC
    `);

    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Server Error" });
  }
});
// POST a new renewal
/*router.post("/", async (req, res) => {
  try {
    const {
      contract_id,
      renewal_date,
      reminder_days,
      status,
    } = req.body;

    const result = await pool.query(
      `INSERT INTO renewals
      (contract_id, renewal_date, reminder_days, status)
      VALUES ($1, $2, $3, $4)
      RETURNING *`,
      [contract_id, renewal_date, reminder_days, status]
    );

    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Server Error" });
  }
});

module.exports = router;*/
router.post("/", async (req, res) => {
  try {
    const {
      contract_id,
      renewal_type,
      renewal_date,
      reminder_days,
      status,
    } = req.body;

    const result = await pool.query(
      `INSERT INTO renewals
      (
        contract_id,
        renewal_type,
        renewal_date,
        reminder_days,
        status
      )
      VALUES ($1, $2, $3, $4, $5)
      RETURNING *`,
      [
        contract_id,
        renewal_type,
        renewal_date,
        reminder_days,
        status,
      ]
    );

    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error("ERROR:", err);
    console.error("MESSAGE:", err.message);
    console.error("DETAIL:", err.detail);
    console.error("CODE:", err.code);

    res.status(500).json({
      message: err.message,
      detail: err.detail,
    });
  }
});

module.exports = router;
// UPDATE renewal
router.put("/:id", async (req, res) => {
  try {
    const { id } = req.params;

    const {
      renewal_type,
      renewal_date,
      reminder_days,
      status,
    } = req.body;

    const result = await pool.query(
      `
      UPDATE renewals
      SET
        renewal_type=$1,
        renewal_date=$2,
        reminder_days=$3,
        status=$4
      WHERE id=$5
      RETURNING *
      `,
      [
        renewal_type,
        renewal_date,
        reminder_days,
        status,
        id,
      ]
    );

    res.json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({
      message: err.message,
    });
  }
});

File: server/api/renewals.py
from fastapi import APIRouter, HTTPException
from src.database.core import get_connection

router = APIRouter()


# GET ALL RENEWALS
@router.get("/")
def get_renewals():

    conn = get_connection()
    cur = conn.cursor()

    try:

        cur.execute("""
            SELECT
                renewals.id,
                contracts.contract_name,
                contracts.client_name,
                renewals.renewal_type,
                renewals.renewal_date,
                renewals.reminder_days,
                renewals.status

            FROM renewals

            JOIN contracts
            ON renewals.contract_id = contracts.id

            ORDER BY renewals.id ASC
        """)

        rows = cur.fetchall()

        renewals = []

        for row in rows:

            renewals.append({

                "id": row[0],
                "contract_name": row[1],
                "client_name": row[2],
                "renewal_type": row[3],
                "renewal_date": row[4],
                "reminder_days": row[5],
                "status": row[6]

            })

        return renewals

    except Exception as e:

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )

    finally:

        cur.close()
        conn.close()
        # ADD NEW RENEWAL
@router.post("/")
def add_renewal(renewal: dict):

    conn = get_connection()
    cur = conn.cursor()

    try:

        cur.execute(
            """
            INSERT INTO renewals
            (
                contract_id,
                renewal_type,
                renewal_date,
                reminder_days,
                status
            )

            VALUES (%s, %s, %s, %s, %s)

            RETURNING *
            """,
            (
                renewal["contract_id"],
                renewal["renewal_type"],
                renewal["renewal_date"],
                renewal["reminder_days"],
                renewal["status"],
            )
        )

        conn.commit()

        row = cur.fetchone()

        return {
            "id": row[0],
            "contract_id": row[1],
            "renewal_date": row[2],
            "reminder_days": row[3],
            "status": row[4],
            "renewal_type": row[5]
        }

    except Exception as e:

        conn.rollback()

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )

    finally:

        cur.close()
        conn.close()
        # UPDATE RENEWAL
@router.put("/{id}")
def update_renewal(id: int, renewal: dict):

    conn = get_connection()
    cur = conn.cursor()

    try:

        cur.execute(
            """
            UPDATE renewals
            SET
                renewal_type = %s,
                renewal_date = %s,
                reminder_days = %s,
                status = %s
            WHERE id = %s
            RETURNING *
            """,
            (
                renewal["renewal_type"],
                renewal["renewal_date"],
                renewal["reminder_days"],
                renewal["status"],
                id
            )
        )

        conn.commit()

        row = cur.fetchone()

        if row is None:
            raise HTTPException(
                status_code=404,
                detail="Renewal not found"
            )

        return {
            "id": row[0],
            "contract_id": row[1],
            "renewal_date": row[2],
            "reminder_days": row[3],
            "status": row[4],
            "renewal_type": row[5]
        }

    except HTTPException:
        raise

    except Exception as e:

        conn.rollback()

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )

    finally:

        cur.close()
        conn.close()